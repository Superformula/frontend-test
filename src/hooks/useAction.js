import { useContext } from 'react';
import { StoreContext } from '../store';

export const useAction = action => {
  if (typeof action !== 'function') {
    console.error(`[action] must be a function`);
    return;
  }

  const [state, dispatch] = useContext(StoreContext);
  const setState = (path, payload) =>
    dispatch({ type: 'SET_STATE', path, payload });

  return (...args) => action(...args)?.({ state, setState });
};
