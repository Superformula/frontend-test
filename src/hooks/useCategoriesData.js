import { useMemo } from 'react';
import { useQuery } from '@apollo/client';
import { categoriesQuery } from '../graphql/queries/categories';

// Mock
import data from '../graphql/mocks/categories.json';

export const useCategoriesData = () => {
  // const { data } = useQuery(categoriesQuery, {
  //   notifyOnNetworkStatusChange: true,
  // });
  const { category = [] } = data?.categories || {};
  const restaurantCategories = category.filter(keepOnlyRestaurants);
  const categories = useMemo(() => reshapeData(restaurantCategories), [
    data?.categories?.category,
  ]);
  return categories;
};

const keepOnlyRestaurants = ({ parent_categories }) =>
  parent_categories?.some(({ alias }) => alias === 'restaurants');

const reshapeData = categories =>
  categories?.map(({ alias, title }) => ({ label: title, value: alias }));
