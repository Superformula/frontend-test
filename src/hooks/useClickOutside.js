import { useState, useEffect } from 'react'

const isClickOutside = (el, clickedElement, setIsOutside) => {
  let targetElement = clickedElement

  do {
    if (targetElement === el) {
      // This is a click inside el
      setIsOutside(false)
    }
    // Go up the DOM
    targetElement = targetElement.parentNode
  } while (targetElement)

  // This is a click outside el
  setIsOutside(true)
}

/**
 * To detect a click outside an HTMLElement (el) we must add a
 * listener on the document element.
 * Then we loop traverse the DOM up from the clicked element to find whether
 * an ancestor of the clicked element belongs to el.
 * @param {HTMLElement} el
 * @param {boolean} active Enable/disable the listener
 */
export default function useClickOutside (el, active) {
  const [isOutside, setIsOutside] = useState(active)

  useEffect(() => {
    const handler = e => isClickOutside(el, e.target, setIsOutside)
    const add = () => document.addEventListener('click', handler)
    const remove = () => document.removeEventListener('click', handler)

    if (el && active) {
      add()
    } else {
      remove()
      setIsOutside(false)
    }

    return remove
  }, [el, active])

  return isOutside
}
