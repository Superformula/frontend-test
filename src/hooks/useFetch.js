import { useState, useEffect } from 'react'
import axios from 'axios'

const cache = {}

const apiState = {
  LOADING: 'LOADING',
  SUCCESS: 'SUCCESS',
  ERROR: 'ERROR'
}

const initialState = {
  state: apiState.LOADING,
  error: null,
  data: null
}

function useFetch (url) {
  const [response, setResponse] = useState(initialState)

  useEffect(() => {
    if (!url) return

    // Create variable for cancel token
    let source = null

    if (cache[url]) {
      setResponse({
        state: apiState.SUCCESS,
        error: null,
        data: cache[url]
      })
    } else {
      // Create cancel token source
      source = axios.CancelToken.source()

      setResponse(initialState)

      axios
        .get(url, {
          // Set cancel token
          cancelToken: source.token
        })
        .then(res => {
          // Cache data
          cache[url] = res.data

          setResponse({
            state: apiState.SUCCESS,
            error: null,
            data: res.data
          })
        })
        .catch(error => {
          if (!axios.isCancel(error)) {
            setResponse({
              state: apiState.ERROR,
              error: error.response ? error.response : true,
              data: null
            })
          }
        })
    }

    return () => {
      // Cancel request if needed
      if (source) {
        source.cancel('Request canceled')
      }
    }
  }, [url])

  return response
}

export default useFetch
