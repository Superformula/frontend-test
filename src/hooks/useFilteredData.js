import { useContext } from 'react';
import { applyMultipleFilters } from 'utils/array';
import { StoreContext } from '../store';

export const useFilteredData = restaurants => {
  const [state] = useContext(StoreContext) || [];
  const { filters = {} } = state || {};
  return applyMultipleFilters(restaurants, filters);
};
