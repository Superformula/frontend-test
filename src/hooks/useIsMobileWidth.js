import { useState, useEffect } from 'react';
import { BREAKPOINTS } from 'consts/screen';

export const useIsMobileWidth = (size = `${BREAKPOINTS.MD}px`) => {
  const [isMobile, setIsMobile] = useState(null);

  const handleResize = () => {
    const isMobileCheck = checkMobileWitdh(`(max-width: ${size})`);
    setIsMobile(isMobileCheck);
  };

  useEffect(() => {
    handleResize();
    window.addEventListener('resize', handleResize);
    return () => window.addEventListener('resize', handleResize);
  }, []);

  return isMobile;
};

const checkMobileWitdh = size => window.matchMedia(size).matches;
