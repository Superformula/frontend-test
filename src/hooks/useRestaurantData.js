import { useMemo } from 'react';
import { useQuery } from '@apollo/client';
import { restaurantQuery } from '../graphql/queries/restaurant';

export const useRestaurantData = id => {
  const { data, loading } = useQuery(restaurantQuery, { variables: { id } });
  const restaurant = useMemo(() => reshapeData(data?.business || {}), [
    data?.business,
  ]);
  console.log({ restaurant });
  return { restaurant, loading };
};

const reshapeData = ({
  categories,
  is_closed,
  location,
  name,
  photos,
  price,
  rating,
  review_count,
  reviews,
}) => ({
  address: location?.formatted_address,
  isOpen: !is_closed,
  name,
  photos,
  price,
  rating,
  reviewCount: review_count,
  type: categories?.[0]?.title,
  reviews: reviews?.map(buildReviews),
});

const buildReviews = ({ rating, time_created, text, user }) => ({
  rating,
  date: new Date(time_created).toLocaleDateString(),
  text,
  user: { name: user.name, avatar: user.image_url },
});
