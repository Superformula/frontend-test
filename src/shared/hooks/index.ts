export * from './usePortal/usePortal';
export * from './useOnClickOutside/useOnClickOutside';
