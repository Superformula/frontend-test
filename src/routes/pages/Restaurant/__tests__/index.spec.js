import React from 'react';
import { shallow } from 'enzyme';
import { useRestaurantData } from 'hooks/useRestaurantData';
import { Restaurant } from '../';

jest.mock('hooks/useRestaurantData');
jest.mock('react-router-dom', () => ({ useParams: jest.fn() }));

describe('Restaurant View', () => {
  it('should render without failing', () => {
    useRestaurantData.mockReturnValueOnce({});
    const wrapper = shallow(<Restaurant />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with loading sections', () => {
    useRestaurantData.mockReturnValueOnce({ loading: true });
    const wrapper = shallow(<Restaurant />);
    expect(wrapper).toMatchSnapshot();
  });
});
