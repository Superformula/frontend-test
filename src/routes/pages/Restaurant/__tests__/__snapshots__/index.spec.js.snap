// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Restaurant View should render with loading sections 1`] = `
<Fragment>
  <GridWrapper>
    <H1>
      <SkeletonLoader />
    </H1>
    <Paragraph
      bottomSpaced={true}
    >
      <SkeletonLoader />
    </Paragraph>
  </GridWrapper>
</Fragment>
`;

exports[`Restaurant View should render without failing 1`] = `
<Fragment>
  <GridWrapper>
    <H1 />
    <Rating
      $lg={true}
      $xs={false}
      max={5}
      rating={0}
    />
    <SpaceBetween
      spaced={true}
    >
      <Paragraph>
         â€¢ 
      </Paragraph>
      <StatusIndicator
        uppercase={null}
        xl={true}
      />
    </SpaceBetween>
  </GridWrapper>
  <LocationImages />
</Fragment>
`;
